pkg/provisioner/ironic/raid.go

```diff
@@ -0,0 +1,172 @@
+
+	// Build target for RAID configuration steps
+	logicalDisks, err = buildTargetRAIDCfg(p.host.Status.Provisioning.RAID)
+	if len(logicalDisks) == 0 || err != nil {
+		return
+	}
+
+	// set root volume
+	if p.host.Status.Provisioning.RootDeviceHints == nil {
+		logicalDisks[0].IsRootVolume = new(bool)
+		*logicalDisks[0].IsRootVolume = true
+	} else {
+		p.log.Info("rootDeviceHints is used, the first volume of raid will not be set to root")
+	}
+
+	// Set target for RAID configuration steps
+	return nodes.SetRAIDConfig(
+		p.client,
+		ironicNode.UUID,
+		nodes.RAIDConfigOpts{LogicalDisks: logicalDisks},
+	).ExtractErr()
+}
+
+// A private method to build RAID disks
+func buildTargetRAIDCfg(raid *metal3v1alpha1.RAIDConfig) (logicalDisks []nodes.LogicalDisk, err error) {
+	if raid == nil {
+		return
+	}
+
+	// build logicalDisks
+	if len(raid.HardwareRAIDVolumes) != 0 {
+		logicalDisks, err = buildTargetHardwareRAIDCfg(raid.HardwareRAIDVolumes)
+	} else if len(raid.SoftwareRAIDVolumes) != 0 {
+		logicalDisks, err = buildTargetSoftwareRAIDCfg(raid.SoftwareRAIDVolumes)
+	}
+
+	return
+}
+
+// A private method to build hardware RAID disks
+func buildTargetHardwareRAIDCfg(volumes []metal3v1alpha1.HardwareRAIDVolume) (logicalDisks []nodes.LogicalDisk, err error) {
+	var (
+		logicalDisk    nodes.LogicalDisk
+		nameCheckFlags map[string]struct{} = make(map[string]struct{})
+	)
+
+	if len(volumes) == 0 {
+		return
+	}
+
+	for _, volume := range volumes {
+		// Check volume's name
+		if volume.Name != "" {
+			_, exist := nameCheckFlags[volume.Name]
+			if exist {
+				return nil, errors.Errorf("name of volume must be unique")
+			}
+			nameCheckFlags[volume.Name] = struct{}{}
+		}
+		// Build logicalDisk
+		logicalDisk = nodes.LogicalDisk{
+			SizeGB:     volume.SizeGibibytes,
+			RAIDLevel:  nodes.RAIDLevel(volume.Level),
+			VolumeName: volume.Name,
+		}
+		if volume.Rotational != nil {
+			if *volume.Rotational {
+				logicalDisk.DiskType = nodes.HDD
+			} else {
+				logicalDisk.DiskType = nodes.SSD
+			}
+		}
+		if volume.NumberOfPhysicalDisks != nil {
+			logicalDisk.NumberOfPhysicalDisks = *volume.NumberOfPhysicalDisks
+		}
+		// Add to logicalDisks
+		logicalDisks = append(logicalDisks, logicalDisk)
+	}
+
+	return
+}
+
+// A private method to build software RAID disks
+func buildTargetSoftwareRAIDCfg(volumes []metal3v1alpha1.SoftwareRAIDVolume) (logicalDisks []nodes.LogicalDisk, err error) {
+	var (
+		logicalDisk nodes.LogicalDisk
+	)
+
+	if len(volumes) == 0 {
+		return
+	}
+
+	if nodes.RAIDLevel(volumes[0].Level) != nodes.RAID1 {
+		return nil, errors.Errorf("the level in first volume of software raid must be RAID1")
+	}
+
+	for _, volume := range volumes {
+		// Build logicalDisk
+		logicalDisk = nodes.LogicalDisk{
+			SizeGB:     volume.SizeGibibytes,
+			RAIDLevel:  nodes.RAIDLevel(volume.Level),
+			Controller: "software",
+		}
+		physicalDisks := makePhysicalDisks(volume.PhysicalDisks)
+		for _, physicalDisk := range physicalDisks {
```

Comment:
Could this loop be replaced by a single assignment if the length of physicalDisks is not 0?

Reply:
Sorry, If we use `logicalDisk.PhysicalDisks = physicalDisks` replace loop, compiler will return `cannot use physicalDisks (variable of type []map[string]string) as []interface{} value in assignment`.